### PR TITLE
feat: add support for timestep bucketing in training

### PR DESCRIPTION
**Pull request by kohya-ss**
This PR introduces timestep bucketing, which reduces the impact of timestep bias when the number of datasets and training epochs are short, such as in LoRA training.

Instead of assigning a timestep to each item in the batch from the range 0-1.0, it does the following:

1. At the beginning of an epoch, prepare the same number of timesteps as the number of data items in one epoch so that the specified number of buckets contain the same number of timesteps. 
2. Concatenate and shuffle them.
3.  When get an item from the dataset, extract the timestep along with it.

Each bucket's timestep has a range of `1000/num_timestep_buckets`, i.e., if `num_timestep_buckets` is 5, it is 0-0.2, 0.2-0.4 ... 0.8-1.0.

For example, if num_timestep_buckets=5 and the total number of data items is 18.

1. 
```
bucket 0 (0.0-0.2): [0.12, 0.15, 0.1, 0.07]
bucket 1 (0.2-0.4): [0.35, 0.22, 0.38, 0.20]
bucket 2 (0.4-0.6): [0.58, 0.42, 0.43, 0.54]
bucket 3 (0.6-0.8): [0.69, 0.73, 0.77, 0.66]
bucket 4 (0.8-1.0): [0.92, 0.89, 0.95, 0.83]
```

2. `[0.73, 0.83, 0.92, 0.15, 0.58, 0.66, 0.69, 0.77, 0.35, 0.89, 0.22, 0.43, 0.07, 0.54, 0.38, 0.12, 0.42, 0.2, 0.1, 0.95]`
-> Truncate to 18 items. `[0.73, 0.83, 0.92, 0.15, 0.58, 0.66, 0.69, 0.77, 0.35, 0.89, 0.22, 0.43, 0.07, 0.54, 0.38, 0.12, 0.42, 0.2]`

```
  --num_timestep_buckets NUM_TIMESTEP_BUCKETS
                        Number of buckets for timestep sampling. Default is None, which disables bucketing. Set to 2
                        or more to enable stratified sampling. This forces timesteps to be sampled uniformly from the
                        [0, 1] range, which can improve training stability, especially for small datasets. /
                        timestepサンプリングのバケット数。デフォルトはNoneで、バケット化を無効にします。2以上に設定する と、層化抽出が有効になり、タイムステップが[0,
                        1]の範囲から均等にサンプリングされるようになります。これは、特に小規模なデータセットでの学習の安定性向上が期待できます。
```

Now `--preserve_distribution_shape` works with, but it doesn't use predetermined random values per epoch, only bucketed timesteps are used.

Also, simultaneous training of high and low in Wan2.2 uses only bucketed timesteps.

### CODE REVIEW

File: src/musubi_tuner/hv_train_network.py [hunk 1]
```diff
@@ -788,57 +841,68 @@ def get_timesteps():
...
                     # https://arxiv.org/abs/2411.14793v3
-                    logsnr = torch.normal(mean=args.logit_mean, std=args.logit_std, size=(batch_size,), device=device)
+                    # logsnr = torch.normal(mean=args.logit_mean, std=args.logit_std, size=(batch_size,), device=device)
+                    logsnr = uniform_to_logsnr_ppF_pytorch(org_timesteps, args.logit_mean, args.logit_std)
                     t = torch.sigmoid(-logsnr / 2)
 
                 elif args.timestep_sampling == "qinglong":
                     # Qinglong triple hybrid sampling: flux_shift:logsnr:logsnr2 = .80:.075:.125
                     # First decide which method to use for each sample independently
                     decision_t = torch.rand((batch_size,), device=device)
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/hv_train_network.py]: `decision_t` is for selecting which sampling method to use, not the value of the time step. It would be meaningful to make this deterministic, but this time we intend to use the time step.

File: src/musubi_tuner/dataset/image_video_dataset.py [hunk 2]
```diff
@@ -653,6 +657,37 @@ def shuffle(self):
...
+            total_timesteps_needed = num_batches * self.batch_size
+
+            # 2. Generate a single large pool of stratified timesteps
+            all_timesteps = []
+            samples_per_bucket = math.ceil(total_timesteps_needed / self.num_timestep_buckets)
+
+            for i in range(self.num_timestep_buckets):
+                min_t = i / self.num_timestep_buckets
+                max_t = (i + 1) / self.num_timestep_buckets
+                for _ in range(samples_per_bucket):
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/dataset/image_video_dataset.py:674]: Doing this would always result in fewer samples in later buckets.

The current code assigns a uniform number of samples to each bucket, then shuffles them and truncates the excess. This guarantees that all buckets end up with the same probability.

File: src/musubi_tuner/hv_train_network.py [hunk 3]
```diff
@@ -1546,11 +1613,17 @@ def train(self, args):
         set_seed(args.seed)
 
         # Load dataset config
+        if args.num_timestep_buckets is not None:
+            logger.info(f"Using timestep bucketing. Number of buckets: {args.num_timestep_buckets}")
+        self.num_timestep_buckets = args.num_timestep_buckets if args.num_timestep_buckets is not None else 0
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/hv_train_network.py]: So it's better to change the checking for values <= 1 only.

File: src/musubi_tuner/hv_train_network.py [hunk 4]
```diff
@@ -749,17 +751,68 @@ def task():
...
+        timesteps = torch.tensor(timesteps, device=device)
+
+        # def uniform_to_normal(t_samples: torch.Tensor) -> torch.Tensor:
+        #     # Clip small values to prevent log(0)
+        #     eps = 1e-7
+        #     t_samples = torch.clamp(t_samples, eps, 1.0 - eps)
+        #     # Convert to logit space with inverse function
+        #     x_samples = torch.log(t_samples / (1.0 - t_samples))
+        #     return x_samples
+
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/hv_train_network.py:793]: I'll leave it with a comment to avoid future changes to the wrong implementation.
